src/main/java/de/tum/cit/aet/artemis/core/service/LLMTokenUsageService.java

```diff
@@ -0,0 +1,76 @@
+package de.tum.cit.aet.artemis.core.service;
+
+import static de.tum.cit.aet.artemis.core.config.Constants.PROFILE_IRIS;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Service;
+
+import de.tum.cit.aet.artemis.core.domain.Course;
+import de.tum.cit.aet.artemis.core.domain.LLMTokenUsage;
+import de.tum.cit.aet.artemis.core.domain.User;
+import de.tum.cit.aet.artemis.core.repository.LLMTokenUsageRepository;
+import de.tum.cit.aet.artemis.exercise.domain.Exercise;
+import de.tum.cit.aet.artemis.iris.domain.message.IrisMessage;
+import de.tum.cit.aet.artemis.iris.service.pyris.dto.data.PyrisLLMCostDTO;
+
+/**
+ * Service for managing the LLMTokenUsage by all LLMs in Artemis
+ */
+@Service
+@Profile(PROFILE_IRIS)
+public class LLMTokenUsageService {
+
+    private final LLMTokenUsageRepository llmTokenUsageRepository;
+
+    public LLMTokenUsageService(LLMTokenUsageRepository llmTokenUsageRepository) {
+        this.llmTokenUsageRepository = llmTokenUsageRepository;
+    }
+
+    /**
+     * saves the tokens used for a specific IrisMessage or Athena call
+     * in case of an Athena call IrisMessage can be null and the
+     * LLMServiceType in tokens has to by Athena
+     *
+     * @param message  IrisMessage related to the TokenUsage
+     * @param exercise Exercise in which the request was made
+     * @param user     User that made the request
+     * @param course   Course in which the request was made
+     * @param tokens   List with Tokens of the PyrisLLMCostDTO Mdel
+     * @return List of the created LLMTokenUsage entries
+     */
+
+    public List<LLMTokenUsage> saveTokenUsage(IrisMessage message, Exercise exercise, User user, Course course, List<PyrisLLMCostDTO> tokens) {
+        List<LLMTokenUsage> tokenUsages = new ArrayList<>();
+
+        // Combine current time and UUID to create a unique traceId
+        long timestamp = System.currentTimeMillis();
+        long uuidComponent = UUID.randomUUID().getLeastSignificantBits() & Long.MAX_VALUE;
+        Long traceId = timestamp + uuidComponent;
```

Comment:
I'd like to see a more deterministic traceId calculation. Currently, if you have multiple `PyrisLLMCostDTO` lists being saved for a single message, they would all have different traceIds, despite belonging to the same final message.
For Iris you could use the id of a PyrisJob to create the traceId, as it will remain the same throughout a job and therefore across multiple status updates for one message.